### ISSUE TITLE
saltpad requires 'develop' and is NOT merged into any 2015.x branches (yet)

### ISSUE THREAD

**Issue by TheBigBear** [tinyclues/saltpad#127]
Dear @Lothiraldan after my recent failed struggles with getting saltpad up and connecting to my master running on 2015.8.3, I have gone back and double checked the salt github repo for the required CORS patches and support that allows saltpad to run.

And what I think I found is very very disappointing. ;-(

As far as I can tell, saltstack HAVE NOT merged the CORS support you provided to any of the current branches? Am I right? It is NOT yet in the 2015.8.x salt-api , right?

It is "ONLY" merged into the 'develop' branch, and NOT as your table sugggests into 2015.8.dev, which I thought meant that the hopefully upcoming 2015.8.4 would then have it. Not so! Please tell me I am wrong on this.

Which means, that UNTIL saltstack decide on a timeline for expanding the CORS support in saltnado and adding this develop support for it officially into the salt-api part of the project officially, saltpad won't fly and will have to wait months - if not years.

Do you know if the additions you made have been considered for the next major release 'Boron'? I can't see any mention of saltstack having put that on their roadmap for that one?

Just to be fair, the saltstack staff did warn and mention this potential danger in the discussions with you when you started helping adding or expanding their existent CORS support in salt-api. (see early warning to you about this situation that I think/understand we are now in here https://github.com/saltstack/salt/issues/26505#issuecomment-136788641 )

I am hoping that I am misreading things and that you know things which I couldn't unearth, but in the meantime, I thought it was imprtant we discussed thisw issue, and once we are sure of the outcome, reflect it in the saltpad Readme.md file appropriately.

**Comment by TheBigBear** [comment]
@whiteinge could you possibly clarify and confirm my findings, or hopefully show/tell me that I am mistaken?

**Comment by Lothiraldan** [comment]
@whiteinge thank you very much for the warning, I think we can kindly ask if we can add improvements made on saltnado for the next 2015.8.x release and in the mean time I will update the README that said that we don't know yet which salt-api version will be compatible with saltpad.

**Comment by whiteinge** [comment]
@Lothiraldan if you open a new pull req that backports your work in saltstack/salt#27142 to the 2015.8 branch and ping me there I will advocate for accepting it because it is a small addition -- even though it is a feature addition. Please triple-check for API regressions and bugs since it will be included in a point-release. I know of at least one very large company that makes use of Saltnado. We can't make a habit of feature additions in point-releases though so don't count on doing this for other additions.

**Comment by whiteinge** [comment]
@Lothiraldan Follow-up question for you. Based on this development and two other questions I've seen in the last five months or so, do you think it would be useful if I were to update saltstack/salt#26505 with more details on the transition to Saltnado? The short version is: in my opinion, at the time of this writing, there is only one reason to use Saltnado over rest_cherrypy (other than personal preference): if you need to handle over one-hundred _concurrent_ requests (the default [thread pool](https://docs.saltstack.com/en/latest/ref/netapi/all/salt.netapi.rest_cherrypy.html#module-salt.netapi.rest_cherrypy.app) setting) then rest_cherrypy cannot do that without heavy memory use because it is a threaded web server but Saltnado can. Excepting for that rest_cherrypy is more mature, more featureful, and more widely deployed. It makes sense to transition to Saltnado for that last-mile performance, because Tornado is pretty dang neat, because there's no reason to have two REST modules with mostly-identical interfaces, and because Salt now deps on Tornado -- but there's a lot of work to be done still (as you know) and I'm not likely to be able to dive in to help out for the foreseeable future. My question to you is: Have I made that clear in the past? And, I'm curious, what is your preference for Saltnado? (I know you ran into trouble with CORS with rest_cherrypy, but I'm still unsure where/how that CORS implementation is broken. Since CORS support is in the current release fixes to it are a perfect fit for point-releases.)

**Comment by whiteinge** [comment]
@TheBigBear Salt releases are [explained here](https://docs.saltstack.com/en/latest/topics/development/git/index.html#release-branching). We create a release branch from the `develop` branch twice a year (IIRC); any features in develop at the time of that creation are the features that will be included in that release and supported (meaning: will see bug fixes) for the lifetime of that release. So it will decidedly not be "years" before new additions to Saltnado make it into a Salt release. The branch for the successor to 2015.8 will _likely_ be created around end of Q1 or beginning of Q2.

**Comment by Lothiraldan** [comment]
@whiteinge I agree about the point that adding features in bug-fixing releases is delicate and I'm not sure if it's worth it, do we have an idea when an hypothetical 2016.Y.Z would be released?

My feeling about saltnado vs rest_cherrypy is that both basically do the conversion between HTTP requests to salt clients call, wait for answer and send back the result. Doing this in a synchronous fashion like rest_cherrypy is very inefficient. While developing saltpad v1, I often get very long api calls (don't remember the number of the thread pool). Combined with the very long blocking calls like getting the entire list of runned jobs, I'm afraid rest_cherrypy will not offer satisfying performance with small to medium load. I don't had particular problems with CORS in rest_cherrypy, more with the general performance of the api.

If we need to make saltnado feature-equivalent to rest_cherrypy to officially recommend saltnado as the preferred netapi, I'm willing to offer help.

@TheBigBear saltpad is still in alpha phase and there is some scenarios where it can hammer the salt-api, is a expected release in Q1 or Q2 blocking you?

**Comment by TheBigBear** [comment]
@Lothiraldan yes, that is a bummer, but looks like it can't be changed, I suppose.  ;-(

I have since switched back to trying to use ```rest_cherrypy``` backed salt-api again, and I now do get in. ;-) just like @virtualguy. (I had a typo in my salt master conf file when it looked as IF saltpad was not able to connect at all.)
 
So I will have a look around as to what works and what doesn't. (Sorry I do KNOW the readme says you expect to run against a ```rest_tornado``` backed ```salt-api```, but I believe, or maybe it is hope in vain, that I will find things that work under a ```rest_cherrypy``` backed ```salt-api``` and do not yet work under ```rest_tornado```, and on that basis hoping to get ```rest_tornado``` fixed (AND Backported to 2015.8.x) to bring it up to feature parity with what a ```rest_cherrypy``` salt-api already supports in the 2015.8.x released branch. I believe this is one of the approaches to the issue that @whiteinge suggested as well?


The first thing I noticed is the 'keys' count does not match the server, but I have seen that you made that explicit now that it is 'fake' data.

Do you happen to know how far I will likely get and where I am most likley going to find that running saltpad v2 against the ```rest_cherrypy``` (NOT rest_tornado) backed version of salt-api?

**Comment by Lothiraldan** [comment]
If CORS support works correctly on rest_cherrypy, you shouldn't hit particular problems. I tried to use the strict minimum base common API (/ and /login), maybe (/jobs), so it should works correctly.

I will also give it a try myself but please give any feedback about rest_cherrypy compatibility.

**Comment by Lothiraldan** [comment]
My bad, the biggest difference will be that you will not have the the real-time updates because rest_cherrypy offer an eventsource endpoint while saltnado offer a websocket endpoint, it shouldn't be too hard to implement on saltpad side if CORS works.

**Comment by TheBigBear** [comment]
@Lothiraldan as far as I can tell the CORS does work in ```rest_cherrypy``` for me, even under ```2015.8.3```, but I'll keep looking and will get back to you, IF and WHEN I find issues. Thanks for all the effort you are putting into ```saltpad``` v2.

**Comment by whiteinge** [comment]
> do we have an idea when an hypothetical 2016.Y.Z would be released?

I'm not sure about that other than late Q1 early Q2. I'll check in with the core team to see if someone there has a better idea.

> Doing this in a synchronous fashion like rest_cherrypy is very inefficient.

Very much agree. Synchronous and HTTP are not a good mix. My preferred usage is to watch the SSE stream since that endpoint is written for long connections. Then to fire async Salt commands (via the `local_async` client; runner and wheel have async options too as of 2014.1 IIRC). And then watch the stream for the eventual returns.

> I'm afraid rest_cherrypy will not offer satisfying performance with small to medium load.

Is this with HTTP or with Websockets? I'm curious because Saltnado and rest_cherrypy should both access Salt's Python API in nearly identical ways. So long as you don't hit the thread-pool limit they should have nearly identical performance.

> If we need to make saltnado feature-equivalent to rest_cherrypy to officially recommend saltnado as the preferred netapi, I'm willing to offer help.

That is indeed the long-term plan. The goal is to be able to have clients like SaltPad and client libraries be able to switch from rest_cherrpy to Saltnado without trouble. Then we can put rest_cherrypy on a deprecation path.

You are more than welcome to dive into any outstanding compatibility issues enumerated in that issue. I would greatly appreciate any help. If you get blocked by anything don't hesitate to ping or email me.

**Comment by whiteinge** [comment]
@TheBigBear thanks for looking into CORS with rest_cherrypy. If you run into any trouble, please open an issue in the Salt repo.

**Comment by Lothiraldan** [comment]
@whiteinge I checked quickly yesterday and the CORS support in rest_cherrypy doesn't seems to work

**Comment by whiteinge** [comment]
If you can point out where it's wrong at the HTTP level I can find a
few minutes to dive into the code. My working CORS knowledge is
limited to the html5rocks tutorial. :-P

**Comment by whiteinge** [comment]
> do we have an idea when an hypothetical 2016.Y.Z would be released?

Around March-ish. Ballpark.

**Comment by Lothiraldan** [comment]
@whiteinge From what I tested, the rest_cherrypy tries to validate the CORS preflight request but the preflight request couldn't includes the "X-Token" header and for the cookie authentication, the rest_cherrypy doesn't set the `Access-Control-Allow-Credentials: true` header (see https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Requests_with_credentials) so it shouldn't works either, but I didn't test the cookie authentication.

**Comment by arj03** [comment]
Agree, ran into the same problem today. 

SaltPad looks really cool, wish it would run out of the box.

**Comment by Lothiraldan** [comment]
@whiteinge As the CORS support is undocumented in rest_cherrypy, if we want to "fix" it, could it be included in next minor release?

**Comment by whiteinge** [comment]
Yeah, that definitely counts as a bug fix. The docs omission was
accidental. (Other endpoints have documented CORS considerations.)

**Comment by TheBigBear** [comment]
@Lothiraldan FYI, the saltstack team have drawn a line in the sand 3 days ago for what code they are going to include in the next 2015.8.x release. 

The tag or (line in the sand) is at https://github.com/saltstack/salt/releases , they typically take about a week from drawing that line until they officially announce and release that new 2015.8.4 . 

I am not entirely sure, what the processes are in between the tagging and the releasing, but I suspect it has to do with creating the various packages for linux and mac and windows distros and testing them all etc, so it wouldn't surprise me if new bug fixes might be put towards the next release after, presumably 2015.8.5? @whiteinge is that the most likely, or do you think fixes at thos stage might still be slipped into the 2015.8.4 release?

At a rough guess, they seem to come out with a new step release roughly on average every two months, judging by the steps in 2015.8.x series so far. And as the branches get older (and presumably more stable), those times or gaps between step releases tend to get a little longer over time.

So I would hazard a guess that having CORS fixes included in the first 2016.x release "Boron", may be still slower than trying actually get them into 2015.8.x, is there anything I can do to help getting a released production version of salt out there that saltpad can run on out of the box? 

And besides, I am pretty sure that, having those saltpad/CORS fixes in the 2015.8.x branch means that at release time of the first 2016.x Boron release those 2015.8.x fixes will be ported forward to the Boron branch before release anyway.

**Comment by Lothiraldan** [comment]
With this PR https://github.com/Lothiraldan/saltpad/pull/142, saltpad is compatible with rest_cherrypy in develop version. If I manage to deploy saltpad via rest_cherrypy (like molten does), it should be compatible with stable rest_cherrypy, I hope.

**Comment by Routhinator** [comment]
Does this mean we'll be able to once again use SaltPad with the current stable salt-api soon? Or are those of us running stable still stuck until the right patches come down the pipeline?

**Comment by Lothiraldan** [comment]
Hi @Routhinator, with the latest version of saltpad, it should works with rest_cherrypy via a special deployment process named "Rest_cherrypy single-app page deployment" (https://github.com/Lothiraldan/saltpad#rest_cherrypy-single-app-page-deployment).
